src/toast/observation_dist.py

```diff
@@ -164,170 +193,285 @@ def __init__(
+            out.append(None)
             continue
-        if pend <= off:
+        if tend <= off:
+            out.append(None)
             continue
-        new_off = off
-        if poff > off:
-            new_off = poff
-        new_n = off + n - new_off
-        if pend < off + n:
-            new_n = pend - new_off
-        pnew.append((ip, new_off - off, new_off, new_n))
-    return pnew
-
-
-def compute_det_sample_offsets(all_dets, old_dist, new_dist):
-    """ """
-    det_order = {y: x for x, y in enumerate(all_dets)}
-
-    old_dets = list()
-    for pdet in old_dist.dets:
-        dfirst = det_order[pdet[0]]
-        dlast = det_order[pdet[-1]]
-        old_dets.append((dfirst, dlast - dfirst + 1))
-
-    new_dets = list()
-    for pdet in new_dist.dets:
-        dfirst = det_order[pdet[0]]
-        dlast = det_order[pdet[-1]]
-        new_dets.append((dfirst, dlast - dfirst + 1))
-
-    # Every process figures out its send and receive information
-
-    send_row = compute_1d(self.local_index_offset, self.n_local_samples, newdist.samps)
-    det_first = det_order[self.local_detectors[0]]
-    det_last = det_order[self.local_detectors[-1]]
-    send_col = compute_1d(det_first, det_last - det_first + 1, newdets)
-    send_info = list()
-    for sc in send_col:
-        for sr in send_row:
-            psend = sc[0] * newdist.comm_row_size + sr[0]
-            send_info.append((psend, sc[1], sc[2], sc[3], sr[1], sr[2], sr[3]))
-
-    msg = "Proc {} det send:  {}".format(self.comm_rank, send_info)
-    print(msg, flush=True)
-
-    recv_row = compute_1d(
-        newdist.samps[newdist.comm_row_rank][0],
-        newdist.samps[newdist.comm_row_rank][1],
-        self.dist.samps,
-    )
-    det_first = det_order[newdist.dets[newdist.comm_col_rank][0]]
-    det_last = det_order[newdist.dets[newdist.comm_col_rank][-1]]
-    recv_col = compute_1d(det_first, det_last - det_first + 1, olddets)
-    recv_info = list()
-    for rc in recv_col:
-        for rr in recv_row:
-            precv = rc[0] * self.dist.comm_row_size + rr[0]
-            recv_info.append((precv, rc[1], rc[2], rc[3], rr[1], rr[2], rr[3]))
-
-    msg = "Proc {} det recv:  {}".format(self.comm_rank, recv_info)
-    print(msg, flush=True)
-
-
-def redistribute(self, process_rows, times=None):
-    """Take the currently allocated observation and redistribute in place.
-
-    This changes the data distribution within the observation.  After
-    re-assigning all detectors and samples, the currently allocated shared data
-    objects and detector data objects are redistributed using the observation
-    communicator.
+        local_off = 0
+        target_off = 0
+        if toff >= off:
+            local_off = toff - off
+        else:
+            target_off = off - toff
+        tnum = n - local_off
+        if tend < off + n:
+            tnum = tend - local_off
+        out.append(slice(local_off, local_off + tnum, 1))
+    return out
+
+
+def redistribute_buffer(
```

Comment:
This routine is 350 lines long and could really do with some refactoring. Not critical, of course.

Reply:
It seems like this comment is also misplaced in the source...  The `redistribute_buffer` function is 120 lines, including docstring and arguments.  Perhaps you mean the `redistribute_data` function?

Reply:
Correct. This comment concerns `redistribute_data`.

Reply:
This is now split into several functions that are called once.  They exist only to keep the length of functions shorter.